### PR TITLE
Work some 'scoped' issues on IDE

### PR DESCRIPTION
**Pull request by Youssef1313**
Related to: #64906

Fixes https://github.com/dotnet/roslyn/issues/65403

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/ScopedKeywordRecommender.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                return true;
+            }
+
+            var token = context.TargetToken;
+            switch (token.Kind())
+            {
+                //     for (scoped ref var x ...
+                case SyntaxKind.OpenParenToken:
+                    var previous = token.GetPreviousToken(includeSkipped: true);
+                    return previous.IsKind(SyntaxKind.ForKeyword);
```
Comment (Youssef1313) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/ScopedKeywordRecommender.cs:45]: This should work with `foreach` as well. I'll fix it and update the tests.

**Comment by CyrusNajmabadi** [comment]
@cston `scoped` is still in preview right?

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/TypeImportCompletionProviderTests.cs [hunk 2]
```diff
@@ -1798,6 +1798,74 @@ namespace Foo
                     inlineDescription: "Foo");
         }
 
+        [WpfFact]
+        public async Task TestAfterScoped()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/TypeImportCompletionProviderTests.cs:1802]: please don't write 'completion provider' tests (note: KeywordCOmpletionProvider is fine).  Please put completion tests in CSharpCompletionCommandHandlerTests instead.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/GlobalKeywordRecommender.cs [hunk 3]
```diff
@@ -34,6 +34,7 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
             return
                 context.IsStatementContext ||
                 context.IsGlobalStatementContext ||
+                context.IsTypeContext ||
                 UsingKeywordRecommender.IsUsingDirectiveContext(context, forGlobalKeyword: true, cancellationToken) ||
                 context.IsAnyExpressionContext ||
                 context.IsObjectCreationTypeContext ||
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/GlobalKeywordRecommender.cs:40]: can likely remove IsIsOrAsTypeContext right?  And IsFunctionPointerTypeContext and IsObjectCreationTypeContext , right?

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/InKeywordRecommender.cs [hunk 4]
```diff
@@ -36,7 +36,7 @@ private static bool IsInParameterModifierContext(int position, CSharpSyntaxConte
             if (context.SyntaxTree.IsParameterModifierContext(
                     position, context.LeftToken, includeOperators: true, out var parameterIndex, out var previousModifier))
             {
-                if (previousModifier == SyntaxKind.None)
+                if (previousModifier is SyntaxKind.None or SyntaxKind.ScopedKeyword)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/InKeywordRecommender.cs:39]: this feels odd.  `IsParameterModifierContext` should understand `Scoped` itself right?

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/OutKeywordRecommender.cs [hunk 5]
```diff
@@ -31,7 +31,7 @@ private static bool IsOutParameterModifierContext(int position, CSharpSyntaxCont
         {
             return context.SyntaxTree.IsParameterModifierContext(
                        position, context.LeftToken, includeOperators: false, out _, out var previousModifier) &&
-                   previousModifier == SyntaxKind.None;
+                   previousModifier is SyntaxKind.None or SyntaxKind.ScopedKeyword;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/OutKeywordRecommender.cs:34]: same feedback.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/RefKeywordRecommender.cs [hunk 6]
```diff
@@ -93,7 +93,7 @@ private static bool IsRefParameterModifierContext(int position, CSharpSyntaxCont
             if (context.SyntaxTree.IsParameterModifierContext(
                     position, context.LeftToken, includeOperators: false, out var parameterIndex, out var previousModifier))
             {
-                if (previousModifier == SyntaxKind.None)
+                if (previousModifier is SyntaxKind.None or SyntaxKind.ScopedKeyword)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/RefKeywordRecommender.cs:96]: same.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/InKeywordRecommender.cs (hunk 4 above)
Comment (Youssef1313) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/InKeywordRecommender.cs:39]: @CyrusNajmabadi I'm not sure what you mean?

**Comment by cston** [comment]
> @cston `scoped` is still in preview right?

`scoped` is no longer in preview; it is available in C#11 with `-langversion:11` or later.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/GlobalKeywordRecommender.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/GlobalKeywordRecommender.cs:40]: Maybe, I'll check.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/TypeImportCompletionProviderTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Youssef1313) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/TypeImportCompletionProviderTests.cs]: Moving to `CSharpCompletionCommandHandlerTests` resolved the issue.

Comment (Youssef1313) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/TypeImportCompletionProviderTests.cs]: Thanks @CyrusNajmabadi. That fixed the test failure.

File: src/EditorFeatures/CSharpTest2/Recommendations/ScopedKeywordRecommenderTests.cs [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+        {
+            await VerifyAbsenceAsync(AddInsideMethod(
+@"foreach (ref $$ x"));
+        }
+
+        [Fact]
+        public async Task TestNotInForEachRefLoop2()
+        {
+            await VerifyAbsenceAsync(AddInsideMethod(
+@"foreach (ref v$$ x"));
```
Comment (cston) [inline src/EditorFeatures/CSharpTest2/Recommendations/ScopedKeywordRecommenderTests.cs:216]: >ref v$$ [](http://example.com/codeflow?start=11&length=7)

Should this be `ref s$$`?

File: src/EditorFeatures/CSharpTest2/Recommendations/ScopedKeywordRecommenderTests.cs [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+        {
+            await VerifyAbsenceAsync(AddInsideMethod(
+@"for (ref readonly $$"));
+        }
+
+        [Fact]
+        public async Task TestNotInForRefReadonlyLoop1()
+        {
+            await VerifyAbsenceAsync(AddInsideMethod(
+@"for (ref readonly v$$"));
```
Comment (cston) [inline src/EditorFeatures/CSharpTest2/Recommendations/ScopedKeywordRecommenderTests.cs:251]: >ref readonly v$$ [](http://example.com/codeflow?start=7&length=16)

`ref readonly s$$`?

File: src/EditorFeatures/CSharpTest2/Recommendations/ScopedKeywordRecommenderTests.cs (hunk 7 above)
Comment (Youssef1313) [reply src/EditorFeatures/CSharpTest2/Recommendations/ScopedKeywordRecommenderTests.cs:216]: Nice catch! @cston 
I originally started with `var` tests.

File: src/EditorFeatures/CSharpTest2/Recommendations/ScopedKeywordRecommenderTests.cs (hunk 8 above)
Comment (Youssef1313) [reply src/EditorFeatures/CSharpTest2/Recommendations/ScopedKeywordRecommenderTests.cs]: Fixed

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/GlobalKeywordRecommender.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/GlobalKeywordRecommender.cs]: Fixed.

**Review by cston — APPROVED** [review]

**Comment by jcouv** [comment]
@CyrusNajmabadi Please take another look. Thanks

**Comment by CyrusNajmabadi** [comment]
I can look tomorrow.  Thanks!

File: src/EditorFeatures/CSharpTest/ReassignedVariable/CSharpReassignedVariableTests.cs [hunk 9]
```diff
@@ -500,13 +500,21 @@ public async Task AssignmentThroughRefLocal()
 using System;
 class C
 {
-    void M(ref int [|p|])
+    void M1(ref int [|p|])
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/ReassignedVariable/CSharpReassignedVariableTests.cs:503]: would prefer not changing existing tests.  add new tests please.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/GlobalKeywordRecommender.cs [hunk 10]
```diff
@@ -32,13 +32,9 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
...
 
             return
-                context.IsStatementContext ||
-                context.IsGlobalStatementContext ||
+                context.IsTypeContext ||
                 UsingKeywordRecommender.IsUsingDirectiveContext(context, forGlobalKeyword: true, cancellationToken) ||
                 context.IsAnyExpressionContext ||
-                context.IsObjectCreationTypeContext ||
-                context.IsIsOrAsTypeContext ||
-                context.IsFunctionPointerTypeArgumentContext ||
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/GlobalKeywordRecommender.cs:41]: 👍

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/InKeywordRecommender.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/InKeywordRecommender.cs:39]: oh, misread.  sorry.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/InKeywordRecommender.cs:39]: can you say `this scoped`?

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/ScopedKeywordRecommender.cs [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            }
+
+            var token = context.TargetToken;
+            switch (token.Kind())
+            {
+                // for (scoped ref var x ...
+                // foreach (scoped ...
+                case SyntaxKind.OpenParenToken:
+                    var previous = token.GetPreviousToken(includeSkipped: true);
+                    return previous.Kind() is SyntaxKind.ForKeyword or SyntaxKind.ForEachKeyword;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/ScopedKeywordRecommender.cs:46]:
```suggestion
                    return token.Parent is ForStatementSyntax or CommonForEachStatementSyntax;
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 12]
```diff
@@ -1230,14 +1239,25 @@ public static bool IsPossibleExtensionMethodContext(this SyntaxTree syntaxTree,
...
+            if (token.Kind() is SyntaxKind.OpenParenToken or SyntaxKind.CommaToken)
             {
-                return true;
+                parent = token.Parent;
+            }
+            else if (token.IsKind(SyntaxKind.ScopedKeyword) && token.Parent.IsKind(SyntaxKind.Parameter))
+            {
+                parent = token.Parent.Parent;
+            }
+            else if (token.IsKind(SyntaxKind.IdentifierToken) && token.Text == "scoped" && token.Parent is IdentifierNameSyntax scopedIdentifierName && scopedIdentifierName.Parent.IsKind(SyntaxKind.Parameter))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1251]: feel like there's somet repititino of these checks.  but i don't feel strongly about it.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 13]
```diff
@@ -1844,6 +1865,13 @@ SyntaxKind.ForStatement or
                 return true;
             }
 
+            // scoped |
+            if ((token.IsKind(SyntaxKind.IdentifierToken) && token.Text == "scoped" && token.Parent.IsKind(SyntaxKind.IdentifierName) && token.Parent.Parent!.Kind() is SyntaxKind.VariableDeclaration or SyntaxKind.ExpressionStatement or SyntaxKind.IncompleteMember) ||
+                token.IsKind(SyntaxKind.ScopedKeyword) && token.Parent!.Kind() is SyntaxKind.ScopedType or SyntaxKind.IncompleteMember)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1870]: &&'s and ||'s are kiling me. just break these into two cases.  and explain the different cases :)

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/InKeywordRecommender.cs (hunk 4 above)
Comment (Youssef1313) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/InKeywordRecommender.cs:39]: I think yes. `this scoped` is valid. I'll add tests for `this scoped`, `this ref`, `this in`, `this scoped ref`, and `this scoped in`.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/ScopedKeywordRecommender.cs (hunk 11 above)
Comment (Youssef1313) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/ScopedKeywordRecommender.cs:46]: @CyrusNajmabadi I'm following the same logic as in other keyword recommenders. I think if a change is needed here it should be in all recommenders?

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs (hunk 13 above)
Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1870]: I found cases that are unhandled while addressing this comment, e.g:

```csharp
scoped $$

class C { }
```

and possibly cases that are either unnecessary or might not be covered by tests. I'll look again.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs (hunk 12 above)
Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:1251]: @CyrusNajmabadi Could we leave the refactoring for a separate PR?

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 14]
```diff
@@ -8885,6 +8885,108 @@ public class AA
...
+                state.Workspace.GlobalOptions.SetGlobalOption(New OptionKey(CompletionOptionsStorage.ForceExpandedCompletionIndexCreation), True)
+                state.Workspace.GlobalOptions.SetGlobalOption(
+                    New OptionKey(CompletionOptionsStorage.ShowItemsFromUnimportedNamespaces, LanguageNames.CSharp), True)
+
+                state.SendInvokeCompletionList()
+                Await state.WaitForAsynchronousOperationsAsync()
+                Await state.WaitForUIRenderedAsync()
+
+                ' Make sure expander is selected
+                state.SetCompletionItemExpanderStateAndWaitForUiRenderAsync(isSelected:=True)
```
Comment (Youssef1313) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:8919]:
```suggestion
                Await state.SetCompletionItemExpanderStateAndWaitForUiRenderAsync(isSelected:=True)
```

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 15]
```diff
@@ -8885,6 +8885,108 @@ public class AA
...
```
Comment (Youssef1313) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:8970]:
```suggestion
                Await state.SetCompletionItemExpanderStateAndWaitForUiRenderAsync(isSelected:=True)
```

**Comment by Youssef1313** [comment]
Merging latest main and re-triggering CI because https://github.com/dotnet/roslyn/pull/65407 *might* have an impact on the logic in this PR

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Othan than flaky tests, everything appears to be passing. This can either be merged as is, or revised to make sure there are no dead checks after the parser changes. Unfortunately, I won't be able to validate now if there are dead checks or not.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Can you restart CI here? Thanks!

**Pull request merged**